File: kmp/iosApp/iosApp/NearbyGatorsContent.swift
import Shared
import SwiftUI

class NearbyGatorsContent: ObservableObject {
    private let gatorsService: GatorsService

    private let listener: Listener

    @Published var nearbyGators: [NearbyGator] = []

    init(gatorsService: GatorsService) {
        self.gatorsService = gatorsService
        self.listener = Listener()
        self.listener.content = self
        gatorsService.addListener(listener: listener)
    }

    deinit {
        gatorsService.removeListener(listener: listener)
    }

    private class Listener : GatorsServiceListener {
        weak var content: NearbyGatorsContent? = nil

        func onGators(nearbyGators: [NearbyGator]) {
            content?.nearbyGators = nearbyGators
        }
    }
}

File: composeApp/src/androidMain/kotlin/com/publicobject/gators/app/NearbyGatorsContent.kt
package com.publicobject.gators.app

import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.snapshots.SnapshotStateList
import com.publicobject.gators.GatorsService
import com.publicobject.gators.NearbyGator

class NearbyGatorsContent(
  gatorsService: GatorsService,
) : GatorsService.Listener {
  val list: SnapshotStateList<NearbyGator> = mutableStateListOf()

  init {
    gatorsService.addListener(this)
  }

  override fun onGators(nearbyGators: List<NearbyGator>) {
    list.clear()
    list.addAll(nearbyGators)
  }
}
